libcraft/core/src/dimension.rs

```diff
@@ -0,0 +1,82 @@
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(try_from = "String", into = "&'static str")]
+pub enum Dimension {
+    Overworld,
+    TheNether,
+    TheEnd,
+}
+
+impl Dimension {
+    pub fn dim_id(&self) -> i32 {
+        match self {
+            Self::Overworld => 0,
+            Self::TheNether => -1,
+            Self::TheEnd => 1,
+        }
+    }
+
+    pub fn from_dim_id(id: i32) -> Option<Self> {
+        match id {
+            0 => Some(Self::Overworld),
+            -1 => Some(Self::TheNether),
+            1 => Some(Self::TheEnd),
+            _ => None,
+        }
+    }
+
+    pub fn namespaced_id(&self) -> &'static str {
+        match self {
+            Self::Overworld => "minecraft:overworld",
+            Self::TheNether => "minecraft:the_nether",
+            Self::TheEnd => "minecraft:the_end",
+        }
+    }
+
+    pub fn from_namespaced_id(id: &str) -> Option<Self> {
+        match id {
+            "minecraft:overworld" => Some(Self::Overworld),
+            "minecraft:the_nether" => Some(Self::TheNether),
+            "minecraft:the_end" => Some(Self::TheEnd),
+            _ => None,
+        }
+    }
+}
+
+impl TryFrom<String> for Dimension {
+    type Error = &'static str;
+
+    fn try_from(namespaced_value: String) -> Result<Self, Self::Error> {
+        if let Some(val) = Self::from_namespaced_id(namespaced_value.as_str()) {
+            Ok(val)
+        } else {
+            Err("Unknown dimension namespaced_id.")
+        }
+    }
+}
+
+impl From<Dimension> for &'static str {
+    fn from(value: Dimension) -> Self {
+        value.namespaced_id()
+    }
+}
+
+impl TryFrom<i32> for Dimension {
```

Comment:
I don't think we should have From/Into conversions for numerical IDs as their use is discouraged but if there is a good reason please tell me.

Reply:
Why the use of `From/Into` for numerical IDs is discouraged?

Reply:
I meant that the use of numerical IDs in general is discouraged by Mojang and thus I am not sure a convenience function like From/Into makes sense to implement

Reply:
Oh, I see. That's true but I don't actually see any problem in implementing `Into/From` for them as despite being discouraged they are not forbidden thus they can be used safely by the developer imo. Also, I think @Defman suggested @Masplus add this in a prior review of his.

Reply:
@Defman What's your opinion about this?

Reply:
I did comment on this earlier.
```
Also this is merely a suggestion, I'm not sure if we should implement From or TryFrom<i32> I'm leaning towards not, just wanted to bring up the possibility.
```
The reason why I'm leaning towards not, is because an `i32` is not very descriptive and the conversion should not happen for all `i32`. Also I agree with @Schuwi that we should prefer namespace id's over ids.

I suggest
1. Remove the `TryFrom`
2. Implement the `NameSpace` trait as discussed
3. Maybe Implement `Numerical` trait with `as_numerical(&self) -> i32` and `from_numerical(i32) -> Option<Self>`.

I'm not sold on the `Numerical` trait.

Reply:
I think the `NameSpace` thing should be discussed in another issue and have its own PR so that we will be able to migrate from simple `str` to `impl NameSpace` on all types that use it. I can open that issue for you now if you are ok with it.